modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java

```diff
@@ -1179,13 +1179,17 @@ private <T> CompletableFuture<T> doWithOneOffRaftGroupService(
             RaftGroupService raftGroupService = raftMgr.startRaftGroupService(MetastorageGroupId.INSTANCE, raftClientConfiguration, true);
 
             return action.apply(raftGroupService)
-                    .whenComplete((res, ex) -> {
+                    // This callback should be executed asynchronously due to
+                    // its code might be done under a busyLock of the raftGroupService,
+                    // and so, it results in a deadlock on shutting down the service.
+                    // TODO: https://issues.apache.org/jira/browse/IGNITE-25787
```

Comment:
I hope you will fix this soon, because this looks very weird

Reply:
:ok_hand: